### PR TITLE
Introduce Client side S3EncryptionClient

### PR DESCRIPTION
**Pull request by MatejNedic**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by timothystone-knsl** [comment]
+1 here. I see value as right now I think I'm blocked on using SSE-C in multipart uploads to a bucket.

**Comment by timothystone-knsl** [comment]
@MatejNedic can you resolve the conflicts? This looks like an important feature. I know I need it.

**Comment by MatejNedic** [comment]
Hey @timothystone-knsl ,

since I am not sure when release will be have to sync with other Maintainers. Will try to push this feature till then, you can define this as S3Client Bean on your own and it should work. Since S3AutoConfiguration should check if there is S3Client bean already existing.

**Comment by MatejNedic** [comment]
@maciejwalkowiak for me it is must to make user handle how he wants to store RSA or AES keys since it is security issue. This is why they need to provide implementation. KMS integration is straight forward. I would like to have this in 3.1.2. if you agree will add docs

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java:124]: shouldn't we check here if these beans are available?

```java
if (ClassUtils.isPresent("software.amazon.encryption.s3.S3EncryptionClient", null)
	&& (rsaProvider.getIfAvailable() != null || aesProvider.getIfAvailable() != null)) {
	return s3EncClient(properties, s3ClientBuilder, rsaProvider, aesProvider);
}
```

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java
Comment (MatejNedic) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java:100]: When building a builder twice it will fail due KeyRing which makes sense since this is a secret. Thinking about should we even have it as a bean.

Because of this trying to assert endpoint region is impossible.

File: docs/src/main/asciidoc/s3.adoc
Comment (maciejwalkowiak) [inline docs/src/main/asciidoc/s3.adoc:174]: Let's add more realistic examples, even if it would mean:

```java
// fetch key pair from secure location
```

File: docs/src/main/asciidoc/s3.adoc
Comment (maciejwalkowiak) [inline docs/src/main/asciidoc/s3.adoc:130]: We need an example on how to use/or at least inject the encryption client.

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AesProvider.java
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AesProvider.java:20]: Since it is a public interface, it needs a Javadoc with author, since, and explanation what is it for

File: spring-cloud-aws-dependencies/pom.xml
Comment (maciejwalkowiak) [inline spring-cloud-aws-dependencies/pom.xml:29]: `amazon.encryption.s3` -> `amazon.encryption.s3.version`

**Review by maciejwalkowiak — CHANGES_REQUESTED** [review]

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java:180]: this will throw if `rsaProvider` is not configured. Shouldn't there be a condition on `s3EncrpytionClientBuilder` bean that it will be auto configured if either `rsaProvider` OR `rsaProvider` bean exists?

Comment (MatejNedic) [reply spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java:174]: @maciejwalkowiak introduced `AnyNestedCondition`. Tests already cover it since they are testing all 3 possibilities.

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java:131]: why static?

**Review by maciejwalkowiak — APPROVED** [review]

**Pull request merged**
